lib/codecs/src/encoding/format/gelf.rs

```diff
@@ -0,0 +1,425 @@
+///   in the event that we need to change the behavior to be more relaxed, so that prior versions
+///   of vector will still work.
+///   The exception is that if 'Additional fields' are found to be missing an underscore prefix and
+///   are otherwise valid field names, we prepend the underscore.
+
+static MISSING_FIELD_STR: &str = "LogEvent does not contain required field.";
+
+static INVALID_TYPE_STR: &str = "LogEvent contains a value with an invalid type.";
+
+static INVALID_FIELD_NAME_STR: &str = "LogEvent contains a value with an invalid type.";
+
+/// Config used to build a `GelfSerializer`.
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+pub struct GelfSerializerConfig;
+
+impl GelfSerializerConfig {
+    /// Creates a new `GelfSerializerConfig`.
+    pub const fn new() -> Self {
+        Self
+    }
+
+    /// Build the `GelfSerializer` from this configuration.
+    pub fn build(&self) -> GelfSerializer {
+        GelfSerializer::new()
+    }
+
+    /// The data type of events that are accepted by `GelfSerializer`.
+    pub fn input_type() -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement() -> schema::Requirement {
+        // While technically we support `Value` variants that can't be losslessly serialized to
+        // JSON, we don't want to enforce that limitation to users yet.
+        schema::Requirement::empty()
+    }
+}
+
+/// Serializer that converts an `Event` to bytes using the GELF format.
+/// Spec: https://docs.graylog.org/docs/gelf
+#[derive(Debug, Clone)]
+pub struct GelfSerializer;
+
+impl GelfSerializer {
+    /// Creates a new `GelfSerializer`.
+    pub fn new() -> Self {
+        GelfSerializer
+    }
+
+    /// Encode event and represent it as JSON value.
+    pub fn to_json_value(&self, event: Event) -> Result<serde_json::Value, serde_json::Error> {
+        match event {
+            Event::Log(log) => serde_json::to_value(&log),
+            Event::Metric(_) | Event::Trace(_) => {
+                panic!("GELF Serializer does not support Metric or Trace events.")
+            }
+        }
+    }
+
+    /// Returns Error for invalid type
+    fn err_invalid_type(
+        &self,
+        field: &str,
+        expected_type: &str,
+        actual_type: &str,
+    ) -> vector_core::Result<()> {
+        Err(format!(
+            "{}: field: {} type: {} expected_type: {}",
+            INVALID_TYPE_STR, field, actual_type, expected_type
+        )
+        .into())
+    }
+
+    /// Creates the clone of input LogEvent if needed and returns mutable
+    fn conformed_log_mut<'a>(
+        &self,
+        log: &LogEvent,
+        conformed_log: &'a mut Option<LogEvent>,
+    ) -> &'a mut LogEvent {
+        if conformed_log.is_none() {
+            *conformed_log = Some(log.clone());
+        }
+        conformed_log.as_mut().unwrap()
+    }
```

Comment:
Could you elaborate a bit why this indirection is necessary?

Reply:
Attempted to address in latest commit.